src/niMidi.cpp

```diff
@@ -234,21 +311,58 @@ class NiMidiSurface: public BaseSurface {
 			this->_sendSysex(CMD_TRACK_NAME, 0, numInBank, name);
 			// todo: level meters, volume, pan
 		}
-		// todo: navigate tracks, navigate banks
+		// todo: navigate tracks, navigate banks. NOTE: probably not here
+	}
+
+	void ClearSelected() {
+		// Clear all selected tracks. Copyright (c) 2010 and later Tim Payne (SWS)
+		int iSel = 0;
+		for (int i = 0; i <= GetNumTracks(); i++) // really ALL tracks, hence no use of CSurf_NumTracks
+			GetSetMediaTrackInfo(CSurf_TrackFromID(i, false), "I_SELECTED", &iSel);
+	}
+
+	void _onTrackSelect(unsigned char numInBank) {
+		int id = this->_bankStart + numInBank;
+		if (id > GetNumTracks()) {
+			return;
+		}
+		MediaTrack* track = CSurf_TrackFromID(id, false);
+		int iSel = 1; // "Select"
+		// If we rather wanted to "Toggle" than just "Select" we would use:
+		// int iSel = *(int*)GetSetMediaTrackInfo(track, "I_SELECTED", nullptr) ? 0 : 1; 
+		ClearSelected(); 
+		GetSetMediaTrackInfo(track, "I_SELECTED", &iSel);		
+	}
+
+	void _onBankSelect(signed char value) {
+		// Manually switch the bank visible in Mixer View WITHOUT influencing track selection
+		int newBankStart = this->_bankStart + (value * BANK_NUM_TRACKS);
+		int numTracks = CSurf_NumTracks(false); // If we ever want to show just MCP tracks in KK Mixer View (param) must be (true)
+		if ((newBankStart < 0) || (newBankStart > numTracks)) {
+			return;
+		}
+		this->_bankStart = newBankStart;
+		this->_MixerUpdate();
 	}
 
 	void _onKnobVolumeChange(unsigned char command, signed char value) {
-		int numInBank = command - CMD_KNOB_VOLUME1 + 1;
+		int numInBank = command - CMD_KNOB_VOLUME1;
+		double dvalue = static_cast<double>(value);
 		MediaTrack* track = CSurf_TrackFromID(numInBank + this->_bankStart, false);
-		if (!track) return;
-		CSurf_SetSurfaceVolume(track, CSurf_OnVolumeChange(track, value * 0.1, true), nullptr);
+		if (!track) {
+			return;
+		}
+		CSurf_OnVolumeChange(track, dvalue * 0.01, true); 
 	}
 
 	void _onKnobPanChange(unsigned char command, signed char value) {
-		int numInBank = command - CMD_KNOB_VOLUME1 + 1;
+		int numInBank = command - CMD_KNOB_PAN1;
+		double dvalue = static_cast<double>(value);
 		MediaTrack* track = CSurf_TrackFromID(numInBank + this->_bankStart, false);
-		if (!track) return;
-		CSurf_SetSurfacePan(track, CSurf_OnPanChange(track, value * 1.0, true), nullptr);
+		if (!track) {
+			return;
+		}
+		CSurf_OnPanChange(track, dvalue * 0.001, true);
```

Comment:
I notice you've changed the multiplier here. I think the value returned from Mk2 and A/M-series keyboards is different: Mk2 has finer control. KK is supposed to send a response to the HELLO message indicating what protocol the keyboard is using (which impacts things like this), but mine doesn't seem to. Does yours?

Reply:
I have not tested if there is any difference between protocols. For the multipliers for volume and pan I went by trial and error and tested with both the coarse control as well as fine control (SHIFT button while turning knobs). The best values I could determine for my S88 Mk2 are:
For volume: / 127 which is * 0.007874
For pan: / (127*8) which is * 0.00098425
They are in the latest commit.

Reply:
UPDATE: Both Pan and Volume scaling will require another fix. Volume still requires logarithmic conversion. Probably easiest to go with Reaper API functions for that (conveniently there are functions like DB2SLIDER already available).
While working on the VU meter update I was briefly considering to use some cleverly speed optimized code for these conversions from here (https://graphics.stanford.edu/~seander/bithacks.html#IntegerLog10) as we don't know how efficient the Reaper implementation is.  In my current version this is called at approx 30fps in the Run() callback and needs to convert 8 stereo channels => 16x30 log conversions per second. But then I thought we may as well trust the Reaper API conversion functions. Just use those everywhere. In the context of knob changes speed is obviously irrelevant and by using it also for the VU meters everything stays consistent.